### PR TITLE
Postgres gets less group and more distinct SQL

### PR DESCRIPTION
**Pull request by bleonard**
## Change description

Better SQL for performance.

### Development

- [x] Application changes have been tested appropriately

### Impact

- [ ] Code follows company security practices and guidelines
- [ ] Security impact of change has been considered
- [ ] Performance impact of change has been considered
- [ ] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [ ] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

File: plugins/@grouparoo/postgres/src/lib/table-import/getPropertyValues.ts [hunk 1]
```diff
@@ -21,55 +21,68 @@ export const getPropertyValues: GetPropertyValuesMethod = async ({
...
-  let query = `SELECT ${aggSelect}, "${tablePrimaryKeyCol}" as __pk FROM %I WHERE`;
-  let addAnd = false;
+  const params: Array<any> = [];
+
+  let query = `SELECT`;
+
+  if (aggFunc || isArray) {
+    query += ` "${tablePrimaryKeyCol}" as __pk`;
+  } else {
+    query += ` DISTINCT ON (__pk) "${tablePrimaryKeyCol}" as __pk`;
```
Comment (evantahler) [inline plugins/@grouparoo/postgres/src/lib/table-import/getPropertyValues.ts:71]: Postgres magic - https://www.geekytidbits.com/postgres-distinct-on/

I don't personally have any knowledge of if `DISTINCT ON` is faster or slower than GROUP BY.

**Review by evantahler — COMMENTED** [review]
Assuming this is part of https://www.pivotaltracker.com/story/show/180125788, I'm guessing this PR will gain peers with similar updates for all the table plugins?

I feel like there should be some test here to make sure that the eventual queries produced look like we expect. It's good all the data loading tests we already have pass... but there's got to be some way to prevent a regression in this.  Perhaps just testing that there's no `group by` when selecting the happy-path exact properties from a table would be enough?

**Comment by evantahler** [comment]
I don’t believe Redshift supports `distinct on` queries. This opens up the can of worms avoided until now about potentially replacing/updating some methods for the redshift plugin that may be different than Postgres.  This would be a similar pattern to what @edmundito is doing with MySQL -> ClickHouse.  This would also mean we need to add tests for Redshift for whatever ends up being different. 


* https://stackoverflow.com/questions/35511803/redshift-postgresql-distinct-on-operator 
* https://dba.stackexchange.com/questions/270155/substitute-distinct-on-for-query-transformation-postgresql-to-redshift

File: plugins/@grouparoo/postgres/src/lib/table-import/getPropertyValues.ts (hunk 1 above)
Comment (bleonard) [reply plugins/@grouparoo/postgres/src/lib/table-import/getPropertyValues.ts:71]: I'm fairly sure you can't just GROUP BY `user_id` 

For example, this query (most recent purchase dates) is invalid.

> SELECT "user_id" as __pk, "created_at" FROM demo.purchases WHERE  "state" = 'successful' AND  "user_id" IN ('1','2','3','4') GROUP BY "user_id" ORDER BY "created_at" DESC 

> ERROR:  column "purchases.created_at" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: SELECT "user_id" as __pk, "created_at" FROM demo.purchases W...

So you have to add `created_at` to the `GROUP BY`. This defeats the purpose as they end up not being grouped.
I believe this works in MySQL and, even better, makes the one that's selected the first in the list.

Comment (bleonard) [reply plugins/@grouparoo/postgres/src/lib/table-import/getPropertyValues.ts:71]: It seems something like this is the way to do the above query without Postgres magic
> SELECT * FROM
>   (SELECT  "user_id" as __pk, "created_at", 
>               ROW_NUMBER() OVER (PARTITION BY "user_id" ORDER BY "created_at" DESC) AS __rownum
>        FROM demo.purchases
>        WHERE  "state" = 'successful' AND  "user_id" IN ('1','2','3','4')
>   ) AS __ranked
> WHERE __ranked.__rownum = 1;

Comment (bleonard) [reply plugins/@grouparoo/postgres/src/lib/table-import/getPropertyValues.ts:71]: Ok, doing that in https://github.com/grouparoo/grouparoo/pull/2522

**Comment by bleonard** [comment]
Replaced by https://github.com/grouparoo/grouparoo/pull/2522

**Pull request closed without merging**